### PR TITLE
handle ACDC of ACDC creation

### LINKED ISSUE #8432: Failure creating ACDC of ACDC in testbed
@ticoann as we discussed in the meeting, we can't create an ACDC of another ACDC in wmstats because it fails the request spec validation (though we don't get the error back in wmstats).

You can see the whole traceback here [1], from the logs in vocms0132.

I think the problem is with this PR:
https://github.com/dmwm/WMCore/pull/8394/files#diff-4d2db220d56730a22e5834d40664acc7L86

where we shouldn't have deleted those lines, otherwise we start sending the ACDC2 arguments through the validation, which will fail since we don't know what the grand grand parent request was.

[1]
```
INFO:cherrypy.access:[01/Feb/2018:16:46:38] vocms0132.cern.ch 188.185.78.132 "GET /reqmgr2/data/request?status=aborted&status=aborted-completed&status=force-complete&detail=False HTTP/1.1" 200 OK [data: 3911 in 415 out 191293 us ] [auth: OK "/DC=ch/DC=cern/OU=computers/CN=wmagent/vocms040.cern.ch" "" ] [ref: "" "WMCore.Services.Requests/v001" ]
ERROR:root:About to raise exception <@========== WMException Start ==========@>
Exception Class: WMSpecFactoryException
Message: There are unknown/unsupported arguments in your request spec: [u'DN', u'ReqMgr2Only', u'RequestWorkflow', u'TotalInputFiles', u'TotalInputEvents', u'_id', u'Teams', u'TotalEstimatedJobs', u'TotalInputLumis', u'OutputModulesLFNBases', u'OutputDatasets', u'InitialPriority']
	ModuleName : WMCore.WMSpec.WMWorkloadTools
	MethodName : validateUnknownArgs
	ClassInstance : None
	FileName : /data/srv/beHG1802c/sw/slc7_amd64_gcc630/cms/reqmgr2/1.1.10.pre6/lib/python2.7/site-packages/WMCore/WMSpec/WMWorkloadTools.py
	ClassName : None
	LineNumber : 315
	ErrorNr : 0

Traceback: 

<@---------- WMException End ----------@>
[01/Feb/2018:16:46:48]  Error: Traceback (most recent call last):
  File "/data/srv/beHG1802c/sw/slc7_amd64_gcc630/cms/reqmgr2/1.1.10.pre6/lib/python2.7/site-packages/WMCore/ReqMgr/Service/Request.py", line 190, in validate
    self._validateRequestBase(param, safe, validate_request_create_args)
  File "/data/srv/beHG1802c/sw/slc7_amd64_gcc630/cms/reqmgr2/1.1.10.pre6/lib/python2.7/site-packages/WMCore/ReqMgr/Service/Request.py", line 85, in _validateRequestBase
    workload, r_args = valFunc(args, self.config, self.reqmgr_db_service, param)
  File "/data/srv/beHG1802c/sw/slc7_amd64_gcc630/cms/reqmgr2/1.1.10.pre6/lib/python2.7/site-packages/WMCore/ReqMgr/Utils/Validation.py", line 87, in validate_request_create_args
    workload, request_args = validate_resubmission_create_args(request_args, config, reqmgr_db_service)
  File "/data/srv/beHG1802c/sw/slc7_amd64_gcc630/cms/reqmgr2/1.1.10.pre6/lib/python2.7/site-packages/WMCore/ReqMgr/Utils/Validation.py", line 138, in validate_resubmission_create_args
    workload = spec.factoryWorkloadConstruction(cloned_args["RequestName"], cloned_args)
  File "/data/srv/beHG1802c/sw/slc7_amd64_gcc630/cms/reqmgr2/1.1.10.pre6/lib/python2.7/site-packages/WMCore/WMSpec/StdSpecs/StdBase.py", line 866, in factoryWorkloadConstruction
    self.validateSchema(schema=arguments)
  File "/data/srv/beHG1802c/sw/slc7_amd64_gcc630/cms/reqmgr2/1.1.10.pre6/lib/python2.7/site-packages/WMCore/WMSpec/StdSpecs/Resubmission.py", line 96, in validateSchema
    self.raiseValidationException(str(ex))
  File "/data/srv/beHG1802c/sw/slc7_amd64_gcc630/cms/reqmgr2/1.1.10.pre6/lib/python2.7/site-packages/WMCore/WMSpec/StdSpecs/StdBase.py", line 905, in raiseValidationException
    raise WMSpecFactoryException(message=msg)
WMSpecFactoryException: <@========== WMException Start ==========@>
Exception Class: WMSpecFactoryException
Message: <@========== WMException Start ==========@>
Exception Class: WMSpecFactoryException
Message: There are unknown/unsupported arguments in your request spec: [u'DN', u'ReqMgr2Only', u'RequestWorkflow', u'TotalInputFiles', u'TotalInputEvents', u'_id', u'Teams', u'TotalEstimatedJobs', u'TotalInputLumis', u'OutputModulesLFNBases', u'OutputDatasets', u'InitialPriority']
	ModuleName : WMCore.WMSpec.WMWorkloadTools
	MethodName : validateUnknownArgs
	ClassInstance : None
	FileName : /data/srv/beHG1802c/sw/slc7_amd64_gcc630/cms/reqmgr2/1.1.10.pre6/lib/python2.7/site-packages/WMCore/WMSpec/WMWorkloadTools.py
	ClassName : None
	LineNumber : 315
	ErrorNr : 0

Traceback: 

<@---------- WMException End ----------@>
	ModuleName : WMCore.WMSpec.StdSpecs.StdBase
	MethodName : raiseValidationException
	ClassInstance : None
	FileName : /data/srv/beHG1802c/sw/slc7_amd64_gcc630/cms/reqmgr2/1.1.10.pre6/lib/python2.7/site-packages/WMCore/WMSpec/StdSpecs/StdBase.py
	ClassName : None
	LineNumber : 905
	ErrorNr : 0

Traceback: 
  File "/data/srv/beHG1802c/sw/slc7_amd64_gcc630/cms/reqmgr2/1.1.10.pre6/lib/python2.7/site-packages/WMCore/WMSpec/StdSpecs/Resubmission.py", line 94, in validateSchema
    validateArgumentsCreate(schema, argumentDefinition)

  File "/data/srv/beHG1802c/sw/slc7_amd64_gcc630/cms/reqmgr2/1.1.10.pre6/lib/python2.7/site-packages/WMCore/WMSpec/WMWorkloadTools.py", line 266, in validateArgumentsCreate
    validateUnknownArgs(arguments, argumentDefinition)

  File "/data/srv/beHG1802c/sw/slc7_amd64_gcc630/cms/reqmgr2/1.1.10.pre6/lib/python2.7/site-packages/WMCore/WMSpec/WMWorkloadTools.py", line 315, in validateUnknownArgs
    raise WMSpecFactoryException(msg)

<@---------- WMException End ----------@>

INFO:cherrypy.error:[01/Feb/2018:16:46:48]  Error: Traceback (most recent call last):
  File "/data/srv/beHG1802c/sw/slc7_amd64_gcc630/cms/reqmgr2/1.1.10.pre6/lib/python2.7/site-packages/WMCore/ReqMgr/Service/Request.py", line 190, in validate
    self._validateRequestBase(param, safe, validate_request_create_args)
  File "/data/srv/beHG1802c/sw/slc7_amd64_gcc630/cms/reqmgr2/1.1.10.pre6/lib/python2.7/site-packages/WMCore/ReqMgr/Service/Request.py", line 85, in _validateRequestBase
    workload, r_args = valFunc(args, self.config, self.reqmgr_db_service, param)
  File "/data/srv/beHG1802c/sw/slc7_amd64_gcc630/cms/reqmgr2/1.1.10.pre6/lib/python2.7/site-packages/WMCore/ReqMgr/Utils/Validation.py", line 87, in validate_request_create_args
    workload, request_args = validate_resubmission_create_args(request_args, config, reqmgr_db_service)
  File "/data/srv/beHG1802c/sw/slc7_amd64_gcc630/cms/reqmgr2/1.1.10.pre6/lib/python2.7/site-packages/WMCore/ReqMgr/Utils/Validation.py", line 138, in validate_resubmission_create_args
    workload = spec.factoryWorkloadConstruction(cloned_args["RequestName"], cloned_args)
  File "/data/srv/beHG1802c/sw/slc7_amd64_gcc630/cms/reqmgr2/1.1.10.pre6/lib/python2.7/site-packages/WMCore/WMSpec/StdSpecs/StdBase.py", line 866, in factoryWorkloadConstruction
    self.validateSchema(schema=arguments)
  File "/data/srv/beHG1802c/sw/slc7_amd64_gcc630/cms/reqmgr2/1.1.10.pre6/lib/python2.7/site-packages/WMCore/WMSpec/StdSpecs/Resubmission.py", line 96, in validateSchema
    self.raiseValidationException(str(ex))
  File "/data/srv/beHG1802c/sw/slc7_amd64_gcc630/cms/reqmgr2/1.1.10.pre6/lib/python2.7/site-packages/WMCore/WMSpec/StdSpecs/StdBase.py", line 905, in raiseValidationException
    raise WMSpecFactoryException(message=msg)
WMSpecFactoryException: <@========== WMException Start ==========@>
Exception Class: WMSpecFactoryException
Message: <@========== WMException Start ==========@>
Exception Class: WMSpecFactoryException
Message: There are unknown/unsupported arguments in your request spec: [u'DN', u'ReqMgr2Only', u'RequestWorkflow', u'TotalInputFiles', u'TotalInputEvents', u'_id', u'Teams', u'TotalEstimatedJobs', u'TotalInputLumis', u'OutputModulesLFNBases', u'OutputDatasets', u'InitialPriority']
	ModuleName : WMCore.WMSpec.WMWorkloadTools
	MethodName : validateUnknownArgs
	ClassInstance : None
	FileName : /data/srv/beHG1802c/sw/slc7_amd64_gcc630/cms/reqmgr2/1.1.10.pre6/lib/python2.7/site-packages/WMCore/WMSpec/WMWorkloadTools.py
	ClassName : None
	LineNumber : 315
	ErrorNr : 0

Traceback: 

<@---------- WMException End ----------@>
	ModuleName : WMCore.WMSpec.StdSpecs.StdBase
	MethodName : raiseValidationException
	ClassInstance : None
	FileName : /data/srv/beHG1802c/sw/slc7_amd64_gcc630/cms/reqmgr2/1.1.10.pre6/lib/python2.7/site-packages/WMCore/WMSpec/StdSpecs/StdBase.py
	ClassName : None
	LineNumber : 905
	ErrorNr : 0

Traceback: 
  File "/data/srv/beHG1802c/sw/slc7_amd64_gcc630/cms/reqmgr2/1.1.10.pre6/lib/python2.7/site-packages/WMCore/WMSpec/StdSpecs/Resubmission.py", line 94, in validateSchema
    validateArgumentsCreate(schema, argumentDefinition)

  File "/data/srv/beHG1802c/sw/slc7_amd64_gcc630/cms/reqmgr2/1.1.10.pre6/lib/python2.7/site-packages/WMCore/WMSpec/WMWorkloadTools.py", line 266, in validateArgumentsCreate
    validateUnknownArgs(arguments, argumentDefinition)

  File "/data/srv/beHG1802c/sw/slc7_amd64_gcc630/cms/reqmgr2/1.1.10.pre6/lib/python2.7/site-packages/WMCore/WMSpec/WMWorkloadTools.py", line 315, in validateUnknownArgs
    raise WMSpecFactoryException(msg)

<@---------- WMException End ----------@>

[01/Feb/2018:16:46:48]  SERVER REST ERROR WMCore.ReqMgr.DataStructs.RequestError.InvalidSpecParameterValue 6b799a0dabe980f4040d0603bbaa4296 (Invalid spec parameter value: <@========== WMException Start ==========@>
Exception Class: WMSpecFactoryException
Message: There are unknown/unsupported arguments in your request spec: [u'DN', u'ReqMgr2Only', u'RequestWorkflow', u'TotalInputFiles', u'TotalInputEvents', u'_id', u'Teams', u'TotalEstimatedJobs', u'TotalInputLumis', u'OutputModulesLFNBases', u'OutputDatasets', u'InitialPriority']
	ModuleName : WMCore.WMSpec.WMWorkloadTools
	MethodName : validateUnknownArgs
	ClassInstance : None
	FileName : /data/srv/beHG1802c/sw/slc7_amd64_gcc630/cms/reqmgr2/1.1.10.pre6/lib/python2.7/site-packages/WMCore/WMSpec/WMWorkloadTools.py
	ClassName : None
	LineNumber : 315
	ErrorNr : 0

Traceback: 

<@---------- WMException End ----------@>)
INFO:cherrypy.error:[01/Feb/2018:16:46:48]  SERVER REST ERROR WMCore.ReqMgr.DataStructs.RequestError.InvalidSpecParameterValue 6b799a0dabe980f4040d0603bbaa4296 (Invalid spec parameter value: <@========== WMException Start ==========@>
Exception Class: WMSpecFactoryException
Message: There are unknown/unsupported arguments in your request spec: [u'DN', u'ReqMgr2Only', u'RequestWorkflow', u'TotalInputFiles', u'TotalInputEvents', u'_id', u'Teams', u'TotalEstimatedJobs', u'TotalInputLumis', u'OutputModulesLFNBases', u'OutputDatasets', u'InitialPriority']
	ModuleName : WMCore.WMSpec.WMWorkloadTools
	MethodName : validateUnknownArgs
	ClassInstance : None
	FileName : /data/srv/beHG1802c/sw/slc7_amd64_gcc630/cms/reqmgr2/1.1.10.pre6/lib/python2.7/site-packages/WMCore/WMSpec/WMWorkloadTools.py
	ClassName : None
	LineNumber : 315
	ErrorNr : 0

Traceback: 

<@---------- WMException End ----------@>)
[01/Feb/2018:16:46:48]    Traceback (most recent call last):
INFO:cherrypy.error:[01/Feb/2018:16:46:48]    Traceback (most recent call last):
[01/Feb/2018:16:46:48]      File "/data/srv/beHG1802c/sw/slc7_amd64_gcc630/cms/reqmgr2/1.1.10.pre6/lib/python2.7/site-packages/WMCore/REST/Server.py", line 711, in default
INFO:cherrypy.error:[01/Feb/2018:16:46:48]      File "/data/srv/beHG1802c/sw/slc7_amd64_gcc630/cms/reqmgr2/1.1.10.pre6/lib/python2.7/site-packages/WMCore/REST/Server.py", line 711, in default
[01/Feb/2018:16:46:48]        return self._call(RESTArgs(list(args), kwargs))
INFO:cherrypy.error:[01/Feb/2018:16:46:48]        return self._call(RESTArgs(list(args), kwargs))
[01/Feb/2018:16:46:48]      File "/data/srv/beHG1802c/sw/slc7_amd64_gcc630/cms/reqmgr2/1.1.10.pre6/lib/python2.7/site-packages/WMCore/REST/Server.py", line 783, in _call
INFO:cherrypy.error:[01/Feb/2018:16:46:48]      File "/data/srv/beHG1802c/sw/slc7_amd64_gcc630/cms/reqmgr2/1.1.10.pre6/lib/python2.7/site-packages/WMCore/REST/Server.py", line 783, in _call
[01/Feb/2018:16:46:48]        v(apiobj, request.method, api, param, safe)
INFO:cherrypy.error:[01/Feb/2018:16:46:48]        v(apiobj, request.method, api, param, safe)
[01/Feb/2018:16:46:48]      File "/data/srv/beHG1802c/sw/slc7_amd64_gcc630/cms/reqmgr2/1.1.10.pre6/lib/python2.7/site-packages/WMCore/ReqMgr/Service/Request.py", line 204, in validate
INFO:cherrypy.error:[01/Feb/2018:16:46:48]      File "/data/srv/beHG1802c/sw/slc7_amd64_gcc630/cms/reqmgr2/1.1.10.pre6/lib/python2.7/site-packages/WMCore/ReqMgr/Service/Request.py", line 204, in validate
[01/Feb/2018:16:46:48]        raise InvalidSpecParameterValue(msg)
INFO:cherrypy.error:[01/Feb/2018:16:46:48]        raise InvalidSpecParameterValue(msg)
[01/Feb/2018:16:46:48]    InvalidSpecParameterValue: InvalidSpecParameterValue 6b799a0dabe980f4040d0603bbaa4296 [HTTP 400, APP 1102, MSG u"Invalid spec parameter value: <@========== WMException Start ==========@>\nException Class: WMSpecFactoryException\nMessage: There are unknown/unsupported arguments in your request spec: [u'DN', u'ReqMgr2Only', u'RequestWorkflow', u'TotalInputFiles', u'TotalInputEvents', u'_id', u'Teams', u'TotalEstimatedJobs', u'TotalInputLumis', u'OutputModulesLFNBases', u'OutputDatasets', u'InitialPriority']\n\tModuleName : WMCore.WMSpec.WMWorkloadTools\n\tMethodName : validateUnknownArgs\n\tClassInstance : None\n\tFileName : /data/srv/beHG1802c/sw/slc7_amd64_gcc630/cms/reqmgr2/1.1.10.pre6/lib/python2.7/site-packages/WMCore/WMSpec/WMWorkloadTools.py\n\tClassName : None\n\tLineNumber : 315\n\tErrorNr : 0\n\nTraceback: \n\n<@---------- WMException End ----------@>", INFO None, ERR None]
INFO:cherrypy.error:[01/Feb/2018:16:46:48]    InvalidSpecParameterValue: InvalidSpecParameterValue 6b799a0dabe980f4040d0603bbaa4296 [HTTP 400, APP 1102, MSG u"Invalid spec parameter value: <@========== WMException Start ==========@>\nException Class: WMSpecFactoryException\nMessage: There are unknown/unsupported arguments in your request spec: [u'DN', u'ReqMgr2Only', u'RequestWorkflow', u'TotalInputFiles', u'TotalInputEvents', u'_id', u'Teams', u'TotalEstimatedJobs', u'TotalInputLumis', u'OutputModulesLFNBases', u'OutputDatasets', u'InitialPriority']\n\tModuleName : WMCore.WMSpec.WMWorkloadTools\n\tMethodName : validateUnknownArgs\n\tClassInstance : None\n\tFileName : /data/srv/beHG1802c/sw/slc7_amd64_gcc630/cms/reqmgr2/1.1.10.pre6/lib/python2.7/site-packages/WMCore/WMSpec/WMWorkloadTools.py\n\tClassName : None\n\tLineNumber : 315\n\tErrorNr : 0\n\nTraceback: \n\n<@---------- WMException End ----------@>", INFO None, ERR None]
[01/Feb/2018:16:46:48] vocms0132.cern.ch 2a01:cb15:1e9:7300:1c4:fdd9:c282:94f0 "POST /reqmgr2/data/request HTTP/1.1" 400 Bad Request [data: 5050 in 1461 out 433420 us ] [auth: OK "/DC=ch/DC=cern/OU=Organic Units/OU=Users/CN=amaltaro/CN=718748/CN=Alan Malta Rodrigues" "" ] [ref: "https://cmsweb-testbed.cern.ch/wmstats/index.html" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.12; rv:58.0) Gecko/20100101 Firefox/58.0" ]
```

### PR DESCRIPTION
**Pull request by ticoann**
fixes #8432

### CODE REVIEW

**Comment by ticoann** [comment]
@amaltaro, Alan please review. Still need to be tested

**Comment by ticoann** [comment]
It is tested now.
https://reqmgr2-dev.cern.ch/reqmgr2/fetch?rid=sryu_ACDC_ACDC_RelVal_Premix_sryu_180201_232602_4505

**Comment by ticoann** [comment]
test this please

**Comment by amaltaro** [comment]
test this please

**Comment by amaltaro** [comment]
test this please

File: src/python/WMCore/WMSpec/StdSpecs/Resubmission.py [hunk 1]
```diff
@@ -83,6 +84,14 @@ def validateSchema(self, schema):
         Since we skip the master validation for Resubmission specs, we better have
         some specific validation
         """
+        #TODO: for the legacy code if ACDC is created before the updated. remove in next release (Mar 2018)
+        if schema.get('OriginalRequestType') == 'Resubmission':
+            # we cannot validate such schema
+            return
+
+        if schema.get("ResubmissionCount", 1) > 1:
```
Comment (amaltaro) [inline src/python/WMCore/WMSpec/StdSpecs/Resubmission.py]: I'm not sure we need these lines. This logic follows exactly the same as OriginalRequestType, just in a different way :) The bottom line is, we cannot validate ACDC of ACDC (thus, OriginalRequestType=Resubmission, or even ResubmissionCount > 1)

**Review by amaltaro — COMMENTED** [review]
Let me run some tests first, I think we only needed to add the OriginalRequestType check back to the spec.

**Comment by amaltaro** [comment]
Actually it's not just putting those 3 lines back because now the concept of `OriginalRequestType` is different. With your latest changes, OriginalRequestType is the RequestType of the very first workflow, while before it was the RequestType of the parent workflow (the one being ACDC'ed).

All in all, it wasn't consistent and it isn't yet, since OriginalRequestType and OriginalRequestName are _sort of_ no longer correlated.

It works, so I'm going to merge it and we can revisit it in the future if we want. I also tested updating Memory, Multicore and TimePerEvent and it works too. Thanks Seangchan!

File: src/python/WMCore/WMSpec/StdSpecs/Resubmission.py (hunk 1 above)
Comment (ticoann) [reply src/python/WMCore/WMSpec/StdSpecs/Resubmission.py]: First line (OriginalRequestType=Resubmission) verifies ACDC of ACDC if the ACDC is created before this patch update.
second line (ResubmissionCount > 1) verifies if the ACDC is created after this patch update.

Both cases (ACDC of ACDC) we won't validate the parameters.

**Comment by ticoann** [comment]
> Actually it's not just putting those 3 lines back because now the concept of OriginalRequestType is different. With your latest changes, OriginalRequestType is the RequestType of the very first workflow, while before it was the RequestType of the parent workflow (the one being ACDC'ed).

Yes I think we discussed before and if we keep the concept the patch is a little messy.
In the wrong run, it might be better to change it though.
We can change OriginalRequest -> ParentRequest.

**Comment by amaltaro** [comment]
Agreed!

**Pull request merged**